src/lib/request.ts

```diff
@@ -0,0 +1,60 @@
+export interface RequestError extends Error {
+  code?: number;
+}
+
+export const newRequestError = (
+  message: string,
+  code?: number
+): RequestError => {
+  const error = new Error(message) as RequestError;
+  if (code !== undefined) {
+    error.code = code;
+  }
+  return error;
+};
+
+export type RequestCallback = (
+  error: RequestError | null,
+  response?: any
+) => void;
+
+/**
+ * request makes a XMLHttpRequest and handles network and parsing errors.
+ */
+export const request = (
+  address: string,
+  body: any,
+  callback: RequestCallback
+): void => {
+  const req = new XMLHttpRequest();
+  req.open("POST", address, true);
+  req.setRequestHeader("Accept", "application/json");
+  req.setRequestHeader("Content-Type", "application/json");
+  req.onreadystatechange = () => {
+    if (req.readyState !== XMLHttpRequest.DONE) {
+      return;
+    }
+
+    if (req.status === 0) {
+      callback(newRequestError("connection error", 0));
```

Comment:
I know I did this but `0` as a code means nothing, should probably just omit the code or make it something better